File: jar/src/main/java/jline/api/qsys/Qsys_mg1_fb.java
/**
 * @file M/G/1 queueing system analysis with FB/LAS scheduling
 *
 * @since LINE 3.0
 */
package jline.api.qsys;

import jline.io.Ret;
import jline.util.matrix.Matrix;

public final class Qsys_mg1_fb {
    private Qsys_mg1_fb() {}

    /**
     * Analyzes an M/G/1 queueing system with FB (Feedback/LAS) scheduling.
     */
    public static Ret.qsys_prio qsys_mg1_fb(Matrix lambda, Matrix mu, Matrix cs) {
        double[] lambdaArr = lambda.toArray1D();
        double[] muArr = mu.toArray1D();
        double[] csArr = cs.toArray1D();

        if (lambdaArr.length != muArr.length || lambdaArr.length != csArr.length) {
            throw new IllegalArgumentException("lambda, mu, and cs must have the same length");
        }

        int K = lambdaArr.length;

        for (int i = 0; i < K; i++) {
            if (lambdaArr[i] <= 0.0) throw new IllegalArgumentException("lambda[" + i + "] must be positive");
            if (muArr[i] <= 0.0) throw new IllegalArgumentException("mu[" + i + "] must be positive");
            if (csArr[i] < 0.0) throw new IllegalArgumentException("cs[" + i + "] must be non-negative");
        }

        double[] rhoI = new double[K];
        double rhoTotal = 0.0;
        for (int i = 0; i < K; i++) {
            rhoI[i] = lambdaArr[i] / muArr[i];
            rhoTotal += rhoI[i];
        }
        if (rhoTotal >= 1.0) {
            throw new IllegalStateException("System is unstable: utilization rho = " + rhoTotal + " >= 1");
        }

        boolean allExp = true;
        for (int i = 0; i < K; i++) {
            if (Math.abs(csArr[i] - 1.0) >= 1e-6) {
                allExp = false;
                break;
            }
        }

        double[] W = new double[K];
        if (allExp) {
            // see _kb/03-api-layer.md for rationale
            for (int k = 0; k < K; k++) {
                double muK = muArr[k];
                double xMax = 20.0 / muK;
                int nIntervals = 4000; // even
                double h = xMax / nIntervals;
                double acc = 0.0;
                for (int s = 0; s <= nIntervals; s++) {
                    double x = s * h;
                    double t = fbResponseExp(x, lambdaArr, muArr);
                    double fx = muK * Math.exp(-muK * x);
                    double wgt = (s == 0 || s == nIntervals) ? 1.0 : (s % 2 == 1 ? 4.0 : 2.0);
                    acc += wgt * t * fx;
                }
                W[k] = acc * h / 3.0;
            }
        } else {
            // General-service heuristic (MATLAB qsys_mg1_fb_general):
            // evaluate the conditional response at x = 1/mu(k)
            for (int k = 0; k < K; k++) {
                double x = 1.0 / muArr[k];
                double rhoX = 0.0;
                for (int i = 0; i < K; i++) {
                    double integralFbar;
                    if (Math.abs(csArr[i] - 1.0) < 1e-10) {
                        integralFbar = (1.0 - Math.exp(-muArr[i] * x)) / muArr[i];
                    } else {
                        integralFbar = Math.min(x, 1.0 / muArr[i]);
                    }
                    rhoX += lambdaArr[i] * integralFbar;
                }

                double numerator = 0.0;
                for (int i = 0; i < K; i++) {
                    double integralTFbar;
                    if (Math.abs(csArr[i] - 1.0) < 1e-10) {
                        double muI = muArr[i];
                        integralTFbar = (1.0 - Math.exp(-muI * x) * (1.0 + muI * x)) / (muI * muI);
                    } else {
                        integralTFbar = Math.min(x * x / 2.0, 1.0 / (muArr[i] * muArr[i]));
                    }
                    numerator += lambdaArr[i] * integralTFbar;
                }

                if (rhoX >= 1.0) {
                    W[k] = Double.POSITIVE_INFINITY;
                } else {
                    double waitingTerm = numerator / ((1.0 - rhoX) * (1.0 - rhoX));
                    double serviceTerm = x / (1.0 - rhoX);
                    W[k] = waitingTerm + serviceTerm;
                }
            }
        }

        double Q = 0.0;
        for (int i = 0; i < K; i++) {
            Q += lambdaArr[i] * W[i];
        }
        double rhohat = Q / (1.0 + Q);

        return new Ret.qsys_prio(new Matrix(W), rhohat);
    }

    /**
     * FB conditional mean response time T(x) for exponential job sizes,
     * matching MATLAB compute_fb_response: T(x) = m(x)/(1-rho(x))^2 + x/(1-rho(x))
     * with rho(x), m(x) the truncated load and truncated second-moment term.
     */
    private static double fbResponseExp(double x, double[] lambdaArr, double[] muArr) {
        int K = lambdaArr.length;
        double rhoX = 0.0;
        double numerator = 0.0;
        for (int i = 0; i < K; i++) {
            double muI = muArr[i];
            double intFbar = (1.0 - Math.exp(-muI * x)) / muI;
            rhoX += lambdaArr[i] * intFbar;
            double intTFbar = (1.0 - Math.exp(-muI * x) * (1.0 + muI * x)) / (muI * muI);
            numerator += lambdaArr[i] * intTFbar;
        }
        if (rhoX >= 1.0) {
            return Double.POSITIVE_INFINITY;
        }
        return numerator / ((1.0 - rhoX) * (1.0 - rhoX)) + x / (1.0 - rhoX);
    }
}

File: jar/src/main/kotlin/jline/api/qsys/Qsys_mg1_fb.kt
/**
 * @file M/G/1 queueing system analysis with FB/LAS scheduling
 *
 * Implements analysis of M/G/1 queues with Feedback (FB) scheduling, also known as
 * Least Attained Service (LAS) or Shortest Elapsed Time (SET).
 *
 * Under FB/LAS, the job with the least attained service (smallest age) receives
 * priority. This is an age-based policy where priority depends on how much service
 * a job has received, not its original or remaining size.
 *
 * For FB, the mean response time for a job of size x is:
 *   E[T(x)]^FB = (lambda * integral_0^x t*F_bar(t)dt) / (1-rho_x)^2 + x / (1-rho_x)
 *
 * where rho_x = lambda * integral_0^x F_bar(t)dt
 *
 * FB is "Always Unfair" but approximates SRPT for heavy-tailed distributions
 * and is practical since job sizes need not be known in advance.
 *
 * References:
 *   - A. Wierman and M. Harchol-Balter, "Classifying scheduling policies with
 *     respect to unfairness in an M/GI/1", SIGMETRICS 2003, Section 3.3.
 *   - L. Kleinrock, "Queueing Systems, Volume II: Computer Applications", 1976.
 *
 * @since LINE 3.0
 */
package jline.api.qsys

import jline.io.Ret.qsys_prio
import jline.util.matrix.Matrix
import kotlin.math.exp
import kotlin.math.min

/**
 * Analyzes an M/G/1 queueing system with FB (Feedback/LAS) scheduling.
 *
 * Under FB/LAS, the job with the least attained service gets priority.
 * This is practical when job sizes are unknown since it only requires
 * tracking how much service each job has received.
 *
 * @param lambda Matrix (column vector) of arrival rates per class.
 * @param mu     Matrix (column vector) of service rates per class.
 * @param cs     Matrix (column vector) of coefficients of variation per class.
 * @return qsys_prio containing:
 *   - W (Matrix): Vector of mean response times per class
 *   - rho (Double): Overall system utilization
 * @throws IllegalArgumentException if inputs have different lengths or non-positive values
 * @throws IllegalStateException if system is unstable (rho >= 1)
 */
fun qsys_mg1_fb(lambda: Matrix, mu: Matrix, cs: Matrix): qsys_prio {
    val lambdaArr = lambda.toArray1D()
    val muArr = mu.toArray1D()
    val csArr = cs.toArray1D()

    require(lambdaArr.size == muArr.size && lambdaArr.size == csArr.size) {
        "lambda, mu, and cs must have the same length"
    }

    val K = lambdaArr.size

    for (i in 0 until K) {
        require(lambdaArr[i] > 0.0) { "lambda[$i] must be positive" }
        require(muArr[i] > 0.0) { "mu[$i] must be positive" }
        require(csArr[i] >= 0.0) { "cs[$i] must be non-negative" }
    }

    // Compute utilizations
    val rhoI = DoubleArray(K) { i -> lambdaArr[i] / muArr[i] }
    val rhoTotal = rhoI.sum()

    check(rhoTotal < 1.0) { "System is unstable: utilization rho = $rhoTotal >= 1" }

    // Compute response times using FB formula
    val W = DoubleArray(K)

    for (k in 0 until K) {
        val x = 1.0 / muArr[k]  // Mean service time for this class

        // rho_x = lambda * integral_0^x F_bar(t)dt
        var rhoX = 0.0
        for (i in 0 until K) {
            val integralFbar = if (kotlin.math.abs(csArr[i] - 1.0) < 1e-10) {
                // Exponential: integral_0^x exp(-mu*t)dt = (1 - exp(-mu*x)) / mu
                (1.0 - exp(-muArr[i] * x)) / muArr[i]
            } else {
                // Approximation for non-exponential
                min(x, 1.0 / muArr[i])
            }
            rhoX += lambdaArr[i] * integralFbar
        }

        // Numerator: lambda * integral_0^x t*F_bar(t)dt
        var numerator = 0.0
        for (i in 0 until K) {
            val integralTFbar = if (kotlin.math.abs(csArr[i] - 1.0) < 1e-10) {
                // Exponential: integral_0^x t*exp(-mu*t)dt = (1 - exp(-mu*x)*(1+mu*x)) / mu^2
                val muI = muArr[i]
                (1.0 - exp(-muI * x) * (1.0 + muI * x)) / (muI * muI)
            } else {
                // Approximation for non-exponential
                min(x * x / 2.0, 1.0 / (muArr[i] * muArr[i]))
            }
            numerator += lambdaArr[i] * integralTFbar
        }

        if (rhoX >= 1.0) {
            W[k] = Double.POSITIVE_INFINITY
        } else {
            val waitingTerm = numerator / ((1.0 - rhoX) * (1.0 - rhoX))
            val serviceTerm = x / (1.0 - rhoX)
            W[k] = waitingTerm + serviceTerm
        }
    }

    // Compute rhohat
    var Q = 0.0
    for (i in 0 until K) {
        Q += lambdaArr[i] * W[i]
    }
    val rhohat = Q / (1.0 + Q)

    return qsys_prio(Matrix(W), rhohat)
}

/**
 * Queueing system M/G/1 with FB/LAS scheduling algorithms.
 */
@Suppress("unused")
class QsysMg1FbAlgo {
    companion object {
        // Class documentation marker for Dokka
    }
}
